documentation/minimal_viable_product_description.md

```diff
@@ -0,0 +1,125 @@
+
+### **2. HydraNet PGM Predictions**  
+Integrate and deploy HydraNet models to produce PGM-level predictions. At least one HydraNet model must be fully operational and producing predictions within the system, ensuring its compatibility alongside legacy models.
+
+### **3. CM Predictions from Legacy Models**  
+Produce CM-level predictions from the majority of models implemented in the old pipeline. The models to be included should be referenced from the model [catalogs](https://github.com/prio-data/views_pipeline/tree/main/documentation/catalogs), which track the transition to the new pipeline. Where applicable, unmodified DARTS models should be considered if their classification performance meets the project’s needs. Complex adaptations for CM models should be deprioritized unless they are critical for ensuring overall system functionality and performance.
+
+### **4. Ensembles**  
+Implement a mean ensemble method for aggregating predictions from both PGM and CM-level models. Additionally, deploy a median ensemble as a shadow model to serve as a baseline for comparison and validation purposes.
+
+### **5. Calibration (Temporary Solution)**  
+Utilize Jim’s existing calibration script to temporarily align ensemble predictions at the PGM and CM levels. While this solution will suffice for the MVP, developing a more robust and accurate calibration method will be a top priority post-MVP as Jim has marked the current solution as "hacky as hell".
+
+### **6. Prefect Orchestration**  
+Implement Prefect orchestration to manage the entire pipeline, focusing on data fetching, model training, and forecast generation. The orchestration should ensure smooth execution of these core tasks without the inclusion of monitoring or alert systems at this stage. Prefect workflows should be modular and clearly defined to allow easy integration of future improvements.
+
+### **7. Testing**  
+Finalize all tests currently in development, focusing on critical tests necessary for core pipeline functionality. These include tests for data integrity, model training, and forecast accuracy. Individual developers should prioritize tests that validate the functionality of key pipeline components before the MVP launch.
+
+### **8. Exclusion of Unmaintainable Code**  
+Exclude unmaintainable or high-technical-debt code from the migration process. A review process should be in place to identify which legacy models or components are unsuitable for migration based on maintainability and technical debt.
+
+### **9. Catalog Reference**  
+The MVP should include most models from the old pipeline, with clear references to the existing model catalogs. These catalogs will be used to track model migration progress and ensure that all key models are accounted for during the transition to the new pipeline.
+
```

Comment:
Three things are missing: (1) prediction output, (2) mapping and other visualisation,  and (3) evaluation

1. Prediction output: The MVP has to include output of constituent models and ensemble model predictions to prediction storage and API, for this to replace the old one.
2. Mapping: no need to write new tools to produce prediction maps, but the output has to be in a form that allows using the existing tools. Then we can write better tools post-MVP
3. The MVP specification has to include something on evaluation. I realise this was not well developed in the current pipeline, but there was some functionality. This may already be implemented in W&B and omitted for that reason, but then a brief mention might be useful. If there are good reasons to keep this out of the MVP specification, adding it has to have top priority post-MPV.

Reply:
Angelica's requests for the API including metadata is clearly part of the MVP.

Reply:
Agreed that the outputs should go to the prediction store and API -- usage of flat files and pickles and other things should be eliminated given that we have a very advanced storage and management system for predictions. All of those tools however exist already, so they don't need to be specified. As soon as a data-frame like object exists in memory, it should go to the prediction store.

Mapping should be part of the report generation routines, not part of the pipeline. The main purpose of the pipeline is to generate predictions; it's already an extremely complicated system, with a lot of moving parts. Feature creep will only make it extremely unmaintainable. Whatever mapping needs to be done should be done by loading data from the prediction store, mapping on demand (either using whatever mapping tools we want, geopandas etc.) and printing the maps. If we want infrastructure to do this, we should set aside time to do this, but given the state of current gen mapping tools, this is a quick project (e.g. two weeks for a person).

Evaluation is part of the current pipeline, but I'm not sure architecturally where it should go and what it should target. Infrastructurally, I think the logical point is also the prediction store, since it operates on generated predicitons (and a centralized store of predictions would lead to a centralized "sweep" with a single codebase and trivial orchestration) but that's something we should discuss (all of us doing architectural stuff).